{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len expr_gene should be 10k: 10702\n",
      "------df head--------\n",
      "   SYMBOL  GSM629719  GSM629721  GSM629723  GSM629726  GSM629728  GSM629730  \\\n",
      "0     A2M  10.782603  12.224706  11.727896  11.807600  12.064787  12.170485   \n",
      "1  A4GALT   7.468264   7.076519   7.529791   7.172451   7.073430   6.980553   \n",
      "2    AAAS   7.608572   7.784315   7.958451   8.025649   8.018630   7.375393   \n",
      "3    AACS   8.400528   8.413255   8.387644   8.548757   8.047159   7.682527   \n",
      "4   AAGAB   6.292121   5.599561   6.003853   5.749613   5.339957   6.086971   \n",
      "\n",
      "   GSM629732  GSM629722  GSM629724  GSM629725  GSM629727  GSM629729  GSM629735  \n",
      "0  12.268920  10.222550  10.196768  11.939311  10.366529  10.594714  10.416336  \n",
      "1   6.625028   6.435303   6.828313   7.086718   6.331259   6.960688   6.403040  \n",
      "2   7.705605   7.230928   7.441335   7.459882   7.239895   7.244419   7.352893  \n",
      "3   7.448159   8.127974   8.168758   8.223162   8.095394   8.243136   8.182277  \n",
      "4   5.790569   6.271555   6.119289   6.393994   6.413306   6.314710   6.468349  \n",
      "---------------------\n",
      "Index(['SYMBOL', 'GSM629719', 'GSM629721', 'GSM629723', 'GSM629726',\n",
      "       'GSM629728', 'GSM629730', 'GSM629732', 'GSM629722', 'GSM629724',\n",
      "       'GSM629725', 'GSM629727', 'GSM629729', 'GSM629735'],\n",
      "      dtype='object')\n",
      "idx: 1 col: GSM629719 1\n",
      "idx: 2 col: GSM629721 1\n",
      "idx: 3 col: GSM629723 1\n",
      "idx: 4 col: GSM629726 1\n",
      "idx: 5 col: GSM629728 1\n",
      "idx: 6 col: GSM629730 1\n",
      "idx: 7 col: GSM629732 1\n",
      "idx: 8 col: GSM629722 0\n",
      "idx: 9 col: GSM629724 0\n",
      "idx: 10 col: GSM629725 0\n",
      "idx: 11 col: GSM629727 0\n",
      "idx: 12 col: GSM629729 0\n",
      "idx: 13 col: GSM629735 0\n",
      "healthy_idx: [8, 9, 10, 11, 12, 13]  disease_idx: [1, 2, 3, 4, 5, 6, 7]\n",
      "rows: 10702 cols: 14\n",
      "length ranked genes:  10702\n",
      "rank_me[0:5] [(0, 'A2M', 1.2411552390476217), (1, 'A4GALT', 0.4580706929999998), (2, 'AAAS', 0.45414782395238085), (3, 'AACS', -0.04087471685714128), (4, 'AAGAB', -0.4926799067023806)]\n",
      "largest_first[0:5] [(7449, 'PTGIS', 5.405768901547616), (1878, 'CNN1', 4.708138615571428), (9211, 'TAGLN', 4.634852568619049), (1183, 'C7', 4.53848487859524), (127, 'ACTG2', 4.476552746)] 10702\n",
      "return_me[0:5]: ['PTGIS', 'CNN1', 'TAGLN', 'C7', 'ACTG2'] 10702\n",
      "*********\n",
      "*********\n",
      "processing: KEGG_GLYCOLYSIS_GLUCONEOGENESIS  ES: 61.26\n",
      "processing: KEGG_CITRATE_CYCLE_TCA_CYCLE  ES: 13.47\n",
      "processing: KEGG_PENTOSE_PHOSPHATE_PATHWAY  ES: 110.31\n",
      "processing: KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS  ES: 41.43\n",
      "processing: KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM  ES: 92.53\n",
      "processing: KEGG_GALACTOSE_METABOLISM  ES: 135.58\n",
      "processing: KEGG_ASCORBATE_AND_ALDARATE_METABOLISM  ES: 38.52\n",
      "processing: KEGG_FATTY_ACID_METABOLISM  ES: 11.63\n",
      "processing: KEGG_STEROID_BIOSYNTHESIS  ES: 22.85\n",
      "processing: KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS  ES: 69.61\n",
      "processing: KEGG_STEROID_HORMONE_BIOSYNTHESIS  ES: 47.74\n",
      "processing: KEGG_OXIDATIVE_PHOSPHORYLATION  ES: 144.82\n",
      "processing: KEGG_PURINE_METABOLISM  ES: 86.56\n",
      "processing: KEGG_PYRIMIDINE_METABOLISM  ES: 60.72\n",
      "processing: KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM  ES: 44.68\n",
      "processing: KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM  ES: 113.02\n",
      "processing: KEGG_CYSTEINE_AND_METHIONINE_METABOLISM  ES: 14.94\n",
      "processing: KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION  ES: 14.7\n",
      "processing: KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS  ES: 40.36\n",
      "processing: KEGG_LYSINE_DEGRADATION  ES: 35.76\n",
      "processing: KEGG_ARGININE_AND_PROLINE_METABOLISM  ES: 67.83\n",
      "processing: KEGG_HISTIDINE_METABOLISM  ES: 38.71\n",
      "processing: KEGG_TYROSINE_METABOLISM  ES: 70.11\n",
      "processing: KEGG_PHENYLALANINE_METABOLISM  ES: 81.06\n",
      "processing: KEGG_TRYPTOPHAN_METABOLISM  ES: 61.21\n",
      "processing: KEGG_BETA_ALANINE_METABOLISM  ES: 54.5\n",
      "processing: KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM  ES: 48.4\n",
      "processing: KEGG_SELENOAMINO_ACID_METABOLISM  ES: 42.19\n",
      "processing: KEGG_GLUTATHIONE_METABOLISM  ES: 108.48\n",
      "processing: KEGG_STARCH_AND_SUCROSE_METABOLISM  ES: 50.21\n",
      "processing: KEGG_N_GLYCAN_BIOSYNTHESIS  ES: 77.34\n",
      "processing: KEGG_OTHER_GLYCAN_DEGRADATION  ES: 86.85\n",
      "processing: KEGG_O_GLYCAN_BIOSYNTHESIS  ES: 4.23\n",
      "processing: KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM  ES: 117.6\n",
      "processing: KEGG_GLYCOSAMINOGLYCAN_DEGRADATION  ES: 165.49\n",
      "processing: KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE  ES: 168.86\n",
      "processing: KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE  ES: 104.9\n",
      "processing: KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE  ES: 102.16\n",
      "processing: KEGG_GLYCEROLIPID_METABOLISM  ES: 129.58\n",
      "processing: KEGG_INOSITOL_PHOSPHATE_METABOLISM  ES: 38.8\n",
      "processing: KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS  ES: 28.48\n",
      "processing: KEGG_GLYCEROPHOSPHOLIPID_METABOLISM  ES: 113.48\n",
      "processing: KEGG_ETHER_LIPID_METABOLISM  ES: 46.84\n",
      "processing: KEGG_ARACHIDONIC_ACID_METABOLISM  ES: 127.63\n",
      "processing: KEGG_LINOLEIC_ACID_METABOLISM  ES: 30.15\n",
      "processing: KEGG_ALPHA_LINOLENIC_ACID_METABOLISM  ES: 33.68\n",
      "processing: KEGG_SPHINGOLIPID_METABOLISM  ES: 86.57\n",
      "processing: KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES  ES: 74.57\n",
      "processing: KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES  ES: 109.53\n",
      "processing: KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES  ES: 95.05\n",
      "processing: KEGG_PYRUVATE_METABOLISM  ES: 28.0\n",
      "processing: KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM  ES: 53.49\n",
      "processing: KEGG_PROPANOATE_METABOLISM  ES: 11.99\n",
      "processing: KEGG_BUTANOATE_METABOLISM  ES: 29.5\n",
      "processing: KEGG_ONE_CARBON_POOL_BY_FOLATE  ES: 4.67\n",
      "processing: KEGG_RIBOFLAVIN_METABOLISM  ES: 29.83\n",
      "processing: KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM  ES: 40.39\n",
      "processing: KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS  ES: 5.84\n",
      "processing: KEGG_FOLATE_BIOSYNTHESIS  ES: 43.25\n",
      "processing: KEGG_RETINOL_METABOLISM  ES: 82.28\n",
      "processing: KEGG_LIMONENE_AND_PINENE_DEGRADATION  ES: 65.49\n",
      "processing: KEGG_NITROGEN_METABOLISM  ES: 50.89\n",
      "processing: KEGG_SULFUR_METABOLISM  ES: 159.99\n",
      "processing: KEGG_AMINOACYL_TRNA_BIOSYNTHESIS  ES: 29.04\n",
      "processing: KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450  ES: 98.17\n",
      "processing: KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS  ES: 45.76\n",
      "processing: KEGG_ABC_TRANSPORTERS  ES: 42.09\n",
      "processing: KEGG_RIBOSOME  ES: 193.22\n",
      "processing: KEGG_RNA_DEGRADATION  ES: 20.46\n",
      "processing: KEGG_RNA_POLYMERASE  ES: 104.07\n",
      "processing: KEGG_BASAL_TRANSCRIPTION_FACTORS  ES: 51.1\n",
      "processing: KEGG_DNA_REPLICATION  ES: 16.24\n",
      "processing: KEGG_SPLICEOSOME  ES: 20.6\n",
      "processing: KEGG_PROTEASOME  ES: 67.47\n",
      "processing: KEGG_PROTEIN_EXPORT  ES: 15.18\n",
      "processing: KEGG_PPAR_SIGNALING_PATHWAY  ES: 78.88\n",
      "processing: KEGG_BASE_EXCISION_REPAIR  ES: 55.74\n",
      "processing: KEGG_NUCLEOTIDE_EXCISION_REPAIR  ES: 18.11\n",
      "processing: KEGG_MISMATCH_REPAIR  ES: 9.0\n",
      "processing: KEGG_HOMOLOGOUS_RECOMBINATION  ES: 2.23\n",
      "processing: KEGG_NON_HOMOLOGOUS_END_JOINING  ES: 39.28\n",
      "processing: KEGG_MAPK_SIGNALING_PATHWAY  ES: 162.63\n",
      "processing: KEGG_ERBB_SIGNALING_PATHWAY  ES: 93.78\n",
      "processing: KEGG_CALCIUM_SIGNALING_PATHWAY  ES: 191.92\n",
      "processing: KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION  ES: 286.09\n",
      "processing: KEGG_CHEMOKINE_SIGNALING_PATHWAY  ES: 154.16\n",
      "processing: KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM  ES: 66.96\n",
      "processing: KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION  ES: 316.31\n",
      "processing: KEGG_CELL_CYCLE  ES: 10.06\n",
      "processing: KEGG_OOCYTE_MEIOSIS  ES: 4.8\n",
      "processing: KEGG_P53_SIGNALING_PATHWAY  ES: 35.47\n",
      "processing: KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS  ES: 22.41\n",
      "processing: KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT  ES: 54.06\n",
      "processing: KEGG_REGULATION_OF_AUTOPHAGY  ES: 25.66\n",
      "processing: KEGG_LYSOSOME  ES: 145.99\n",
      "processing: KEGG_ENDOCYTOSIS  ES: 130.8\n",
      "processing: KEGG_PEROXISOME  ES: 29.41\n",
      "processing: KEGG_MTOR_SIGNALING_PATHWAY  ES: 23.01\n",
      "processing: KEGG_APOPTOSIS  ES: 23.62\n",
      "processing: KEGG_CARDIAC_MUSCLE_CONTRACTION  ES: 170.58\n",
      "processing: KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION  ES: 140.5\n",
      "processing: KEGG_WNT_SIGNALING_PATHWAY  ES: 52.15\n",
      "processing: KEGG_DORSO_VENTRAL_AXIS_FORMATION  ES: 80.34\n",
      "processing: KEGG_NOTCH_SIGNALING_PATHWAY  ES: 87.91\n",
      "processing: KEGG_HEDGEHOG_SIGNALING_PATHWAY  ES: 130.75\n",
      "processing: KEGG_TGF_BETA_SIGNALING_PATHWAY  ES: 123.3\n",
      "processing: KEGG_AXON_GUIDANCE  ES: 102.86\n",
      "processing: KEGG_VEGF_SIGNALING_PATHWAY  ES: 118.74\n",
      "processing: KEGG_FOCAL_ADHESION  ES: 222.91\n",
      "processing: KEGG_ECM_RECEPTOR_INTERACTION  ES: 224.69\n",
      "processing: KEGG_CELL_ADHESION_MOLECULES_CAMS  ES: 232.16\n",
      "processing: KEGG_ADHERENS_JUNCTION  ES: 61.08\n",
      "processing: KEGG_TIGHT_JUNCTION  ES: 69.6\n",
      "processing: KEGG_GAP_JUNCTION  ES: 118.35\n",
      "processing: KEGG_COMPLEMENT_AND_COAGULATION_CASCADES  ES: 219.23\n",
      "processing: KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION  ES: 154.5\n",
      "processing: KEGG_RENIN_ANGIOTENSIN_SYSTEM  ES: 125.26\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing: KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY  ES: 78.58\n",
      "processing: KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY  ES: 129.77\n",
      "processing: KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY  ES: 27.41\n",
      "processing: KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY  ES: 113.95\n",
      "processing: KEGG_JAK_STAT_SIGNALING_PATHWAY  ES: 120.09\n",
      "processing: KEGG_HEMATOPOIETIC_CELL_LINEAGE  ES: 222.49\n",
      "processing: KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY  ES: 151.6\n",
      "processing: KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY  ES: 31.09\n",
      "processing: KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY  ES: 67.84\n",
      "processing: KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY  ES: 43.9\n",
      "processing: KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS  ES: 120.74\n",
      "processing: KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION  ES: 132.74\n",
      "processing: KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION  ES: 199.21\n",
      "processing: KEGG_CIRCADIAN_RHYTHM_MAMMAL  ES: 56.69\n",
      "processing: KEGG_LONG_TERM_POTENTIATION  ES: 21.19\n",
      "processing: KEGG_NEUROTROPHIN_SIGNALING_PATHWAY  ES: 66.8\n",
      "processing: KEGG_LONG_TERM_DEPRESSION  ES: 65.26\n",
      "processing: KEGG_OLFACTORY_TRANSDUCTION  ES: 145.43\n",
      "processing: KEGG_TASTE_TRANSDUCTION  ES: 113.38\n",
      "processing: KEGG_REGULATION_OF_ACTIN_CYTOSKELETON  ES: 63.91\n",
      "processing: KEGG_INSULIN_SIGNALING_PATHWAY  ES: 77.5\n",
      "processing: KEGG_GNRH_SIGNALING_PATHWAY  ES: 76.37\n",
      "processing: KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION  ES: 4.32\n",
      "processing: KEGG_MELANOGENESIS  ES: 108.0\n",
      "processing: KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY  ES: 99.28\n",
      "processing: KEGG_PATHWAYS_IN_CANCER  ES: 69.91\n",
      "processing: KEGG_ENDOMETRIAL_CANCER  ES: 26.41\n",
      "processing: KEGG_PRIMARY_IMMUNODEFICIENCY  ES: 119.54\n",
      "q10 max: 316.31\n",
      "*********\n",
      "num genes: 7262 num unique: 3318\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "\n",
    "class GSEA:\n",
    "    def __init(self):\n",
    "        self.df=None\n",
    "        self.samp=None\n",
    "        self.expr_gene=None\n",
    "        self.pathway_gene_dict = None\n",
    "        self.largest=None\n",
    "        self.nt=None\n",
    "        \n",
    "    def load_expression_data(self,file_name,drop):\n",
    "        '''\n",
    "        input: gene expression filename\n",
    "        output: pandas dataframe\n",
    "        '''\n",
    "        df = pd.read_csv(file_name,sep='\\t')\n",
    "        if drop:\n",
    "            df = df.drop(columns='SYMBOL')\n",
    "        #print(df.head())\n",
    "        return df\n",
    "\n",
    "    def get_expression_genes(self,df):\n",
    "        '''\n",
    "        input: dataframe\n",
    "        output: list of genes, 10712\n",
    "        '''\n",
    "        df_small = df.iloc[:,df.columns==\"SYMBOL\"]\n",
    "        #print(df_small.shape)\n",
    "        #print(df_small.head())\n",
    "        gene_list = df_small.to_numpy().flatten()\n",
    "        #print(\"len expression genes:\",len(gene_list),gene_list.shape)\n",
    "        return gene_list \n",
    "\n",
    "    def load_samples(self,file_name):\n",
    "        '''\n",
    "        input: filename of samples\n",
    "        output: distionary, key=patient name, value=0 healthy, 1 diseased\n",
    "        '''\n",
    "        d = {}\n",
    "        with open(file_name, \"r\") as fh:\n",
    "            lines = fh.readlines()\n",
    "            for line in lines:\n",
    "                tokens = line.split()\n",
    "                d[tokens[0]] = tokens[1]\n",
    "        #print(\"d:\",d)\n",
    "        return d\n",
    "    \n",
    "    def load_kegg(self,filename):\n",
    "        '''\n",
    "        input:kegg file name\n",
    "        output: dictionary key=KEGG_CITRATE_CYCLE_TCA_CYCLE value=list of genes which are in expression data\n",
    "        all_genes all genes in all pathways no screening\n",
    "        hum_genes, only count ones in expression file\n",
    "        glist only count ones in expression file. glist/path_gene and hum_genes difference? pathgene dict of pathway, list of genes in expressoin\n",
    "        '''\n",
    "        path_gene = {}\n",
    "        with open(filename, \"r\") as fh:\n",
    "            lines = fh.readlines()\n",
    "            #print(\"len(lines):\",len(lines))\n",
    "            for i in range(0,len(lines)):\n",
    "                tokens = lines[i].split()\n",
    "                #print(\"pathway:\",tokens[0])\n",
    "                genes = tokens[2:]\n",
    "                #if i==0:\n",
    "                    #print(\"lines[i]:\",lines[i])\n",
    "                    #print(\"tokens[2:]\",tokens[2:],len(tokens[2:]))\n",
    "                #print(\"pathway genes:\",genes,\"num pathway genes:\",len(genes))\n",
    "                g_list=[]\n",
    "                for g in genes:\n",
    "                    #print(\"+++++++++++testing g:\",g)\n",
    "                    if g in self.expr_gene:\n",
    "                        #print(\"g is in self.expr_gene!!!!!\")\n",
    "                        g_list.append(g)\n",
    "                        #if gene in expression data add to hum_gene dict\n",
    "                #if i==0:\n",
    "                    #print(\"len g_list, should match later query:\",len(g_list))\n",
    "                path_gene[tokens[0]] = g_list\n",
    "                #path_gene[tokens[0]] =tokens[2:]\n",
    "                #print(\"pathway:\",tokens[0],\" num genes:\",len(g_list))\n",
    "            #print(\" len path_gene:\",len(path_gene))\n",
    "            return path_gene\n",
    "\n",
    "    \n",
    "    def load_data(self,expfile,sampfile,keggfile):\n",
    "        self.samp = self.load_samples(sampfile)\n",
    "        self.df = self.load_expression_data(expfile,False)       \n",
    "        self.expr_gene = self.get_expression_genes(self.df)\n",
    "        print(\"len expr_gene should be 10k:\",len(self.expr_gene))\n",
    "        #print(\"expr_gene:\",self.expr_gene)\n",
    "        self.pathway_gene_dict = self.load_kegg(keggfile)\n",
    "        #print(\"************\")\n",
    "        #print(self.pathway_gene_dict[\"KEGG_CITRATE_CYCLE_TCA_CYCLE\"])\n",
    "        #print(len(self.pathway_gene_dict[\"KEGG_CITRATE_CYCLE_TCA_CYCLE\"]))\n",
    "        #print(\"************\")\n",
    "        #pathway_gene_dict = load_kegg(\"c2.cp.kegg.v6.2.symbols.filtered.gmt\",expr_gene)\n",
    "        \n",
    "\n",
    "    def get_gene_rank_order(self):\n",
    "        print(\"------df head--------\")\n",
    "        print(self.df.head())\n",
    "        print(\"---------------------\")\n",
    "        rows,cols = self.df.shape\n",
    "        #build healthy and diseased indexes\n",
    "        print(self.df.columns)\n",
    "        healthy_idx=[]\n",
    "        disease_idx=[]\n",
    "        for idx,col in enumerate(self.df.columns):\n",
    "            if col!=\"SYMBOL\":\n",
    "                print(\"idx:\",idx,\"col:\",col,self.samp[col])\n",
    "                if self.samp[col]=='1':\n",
    "                    disease_idx.append(idx)\n",
    "                elif self.samp[col]=='0':\n",
    "                    healthy_idx.append(idx)\n",
    "                else:\n",
    "                    print(\"xxxx\")\n",
    "\n",
    "        print(\"healthy_idx:\",healthy_idx,\" disease_idx:\",disease_idx)\n",
    "        df_np = self.df.to_numpy()\n",
    "        print(\"rows:\",rows,\"cols:\",cols)#fetch healthy as list of columns, diseased as list of columns\n",
    "        rank_me=[]\n",
    "        healthy_mean=np.mean(df_np[0,[healthy_idx]])\n",
    "        disease_mean=np.mean(df_np[0,[disease_idx]])\n",
    "        for idx in range(0,df_np.shape[0]):\n",
    "            #print(\"df iloc:\",df.iloc[idx,0],df_np[0,[healthy_idx]],healthy_mean)\n",
    "            #print(df_np[0,[disease_idx]],disease_mean)\n",
    "            #print(\"diff:\",disease_mean-healthy_mean)\n",
    "            healthy_mean=np.mean(df_np[idx,[healthy_idx]])\n",
    "            disease_mean=np.mean(df_np[idx,[disease_idx]])\n",
    "            rank_me.append((idx,self.df.iloc[idx,0],disease_mean-healthy_mean)) #sort by second tuple value\n",
    "\n",
    "        print(\"length ranked genes: \",len(rank_me))\n",
    "        #before sorting\n",
    "        print(\"rank_me[0:5]\",rank_me[0:5])\n",
    "        largest_first = sorted(rank_me, key = lambda x: x[2],reverse=True)\n",
    "        print(\"largest_first[0:5]\",largest_first[0:5],len(largest_first))\n",
    "        #return list of genes only largest first\n",
    "        return_me=[]\n",
    "        for x in largest_first:\n",
    "            return_me.append(x[1])\n",
    "        print(\"return_me[0:5]:\",return_me[0:5],len(return_me))\n",
    "        self.nt=len(return_me)\n",
    "        self.largest = return_me\n",
    "        return return_me\n",
    "    \n",
    "    def get_enrichment_score(self,geneset):\n",
    "        #print(\"num pathways, should be 145:\",len(self.pathway_gene_dict))\n",
    "        #print(\"pathway genes\",self.pathway_gene_dict[geneset],len(self.pathway_gene_dict[geneset]))\n",
    "        \n",
    "        ng = len(self.pathway_gene_dict[geneset])\n",
    "        nt=self.nt\n",
    "        penalty_add =np.sqrt((nt-ng)/ng)\n",
    "        penalty_miss = np.sqrt(ng/(nt-ng))\n",
    "        #print(\"ng:\",ng,\" penalty_add:\",penalty_add,\" penalty_miss:\",penalty_miss)    \n",
    "        #calculate ES for each geneset pathway_gene_dict[x] using ranked list. wow for all 10k? \n",
    "        es=[]\n",
    "        es.append(0.)\n",
    "        sum_es=0.\n",
    "        for idx in range(0,len(self.largest)):\n",
    "            #print(idx,\"processing gene:\",self.largest[idx])\n",
    "            #this is right list? not largest first??? wait the geneset has to be in expressoin data\n",
    "            if self.largest[idx] in self.pathway_gene_dict[geneset]:\n",
    "                sum_es +=penalty_add\n",
    "            else:\n",
    "                sum_es-=penalty_miss\n",
    "            es.append(sum_es)\n",
    "        #print(\"es:\",es)\n",
    "        #print(\"es max:\",max(es),round(max(es),2))\n",
    "        return round(max(es),2)\n",
    "    \n",
    "    def q10(self):\n",
    "        \"\"\"\n",
    "        which geneset has highest ES? KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION\n",
    "        \"\"\"\n",
    "        score=[]\n",
    "        for x in self.pathway_gene_dict.keys():\n",
    "            s = self.get_enrichment_score(x)\n",
    "            score.append(s)\n",
    "            print(\"processing:\",x,\" ES:\",s)\n",
    "        print(\"q10 max:\",max(score))\n",
    "        \n",
    "    def q12(self):\n",
    "        \"\"\"\n",
    "        how many unique genes in kegg file\n",
    "        \"\"\"\n",
    "        g=[]\n",
    "        for x in self.pathway_gene_dict.keys():\n",
    "            genes = self.pathway_gene_dict[x]\n",
    "            #print(\"pathway:\",x,\" genes:\",genes)\n",
    "            g.extend(genes)\n",
    "        set_genes=set(g)\n",
    "        print(\"num genes:\",len(g),\"num unique:\",len(set_genes))\n",
    "        \n",
    "    def histogram(self):\n",
    "        '''\n",
    "        this was done somewhere else already\n",
    "        '''\n",
    "        \n",
    "        \n",
    "        \n",
    "gsea=GSEA()\n",
    "gsea.load_data(\"GSE25628_filtered_expression.txt\",\"GSE25628_samples.txt\",\"c2.cp.kegg.v6.2.symbols.filtered.gmt\")\n",
    "#this mirrors self.largest set in class\n",
    "l = gsea.get_gene_rank_order()\n",
    "print(\"*********\")\n",
    "gsea.get_enrichment_score(\"KEGG_CITRATE_CYCLE_TCA_CYCLE\")\n",
    "print(\"*********\")\n",
    "\n",
    "gsea.q10()\n",
    "print(\"*********\")\n",
    "gsea.q12()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
