{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"images/pa1_quiz_1.png\" >\n",
    "<img src=\"images/pa1_quiz_2.png\" >\n",
    "local\n",
    "<img src=\"images/pa1_quiz_3.png\" >\n",
    "we want common substring between the 2 sequences; much smaller sequence part of reference. \n",
    "<img src=\"images/pa1_quiz_4.png\" >\n",
    "definition:\n",
    "0 0 0 0       ;gap open penalty for A, gap extension for A, open for B, extend B\n",
    "from durbin d is gap open penalty and e is gp extension penalty \n",
    "x is the first sequence, y is the second sequence from durbin.pdf\n",
    "0.7=open gap penalty for A or x is dx\n",
    "0.1 = extension penalty for sequnece A is ex\n",
    "1.5 = open gap penalty for B or y is dy\n",
    "0.5 = extension penalty for sequence B or y. is ey \n",
    "durbin d is the open gap penalty, e is the gap extension penalty. \n",
    "highest 1.5 dy\n",
    "0.7 dx\n",
    "0.5 ey\n",
    "0.1 ex \n",
    "or dy, dx, ey, ex\n",
    "<img src=\"images/pa1_quiz_5.png\" >\n",
    "opening a gap is always higher than extension so dx>ex and ex>ey. The second sequence is DNA. The open gap penalty for DNA higher bc harder to get gap bc double helix structure of DNA compared\n",
    "to RNA. Second sequence is DNA.  RNA is first sequence. (verified)\n",
    "<img src=\"images/pa1_quiz_6.png\" >\n",
    "3\n",
    "<img src=\"images/pa1_quiz_7.png\" >\n",
    "no\n",
    "<img src=\"images/pa1_quiz_8.png\" >\n",
    "not compared to the energy difference of a gap. Mismatch penalty is relative to task. IF there is a higher probability\n",
    "of a gap after a UT match then we can add a penalty. No mismatch penalty means gaps are equally probable for all matches. \n",
    "<img src=\"images/pa1_quiz_9.png\" >\n",
    "2136\n",
    "<img src=\"images/pa1_quiz_10.png\" >\n",
    "<img src=\"images/pa1_quiz_11.png\" >\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h6>Q2</h6>\n",
    "homologs - \n",
    "orthologs - \n",
    "paralogs - \n",
    "<img src=\"images/pa1_quiz_12.png\" >\n",
    "global\n",
    "<img src=\"images/pa1_quiz_13.png\" >\n",
    "compare structural or functional annotations. Homology metrics require a score identity and similarity for helping to determine homology. \n",
    "<img src=\"images/pa1_quiz_14.png\" >\n",
    "Average number of mismatches per alignment: 82.50\n",
    "Average number of gaps per alignment: 20.67\n",
    "Average length of gaps (this is the average of all averages): 1.50\n",
    "('avg_mis:', 82.5, 'avg_num_gaps:', 20.666666666666668, ' avg_gap_len:', 1.504387814229314)\n",
    "<img src=\"images/pa1_quiz_15.png\" >\n",
    "human-chimp:2016.0\n",
    "human-turtle:1787.0\n",
    "human-mouse:1953.0\n",
    "<img src=\"images/pa1_quiz_16.png\" >\n",
    "2016.0\n",
    "<img src=\"images/pa1_quiz_17.png\" >\n",
    "turtle\n",
    "<img src=\"images/pa1_quiz_18.png\" >\n",
    "1787.0\n",
    "<img src=\"images/pa1_quiz_19.png\" >\n",
    "<img src=\"images/pa1_quiz_20.png\" >\n",
    "<img src=\"images/pa1_quiz_21.png\" >"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h6>Q3</h6>\n",
    "\n",
    "\n",
    "<img src=\"images/pa1_quiz_22.png\" >\n",
    "local, this https://prosite.expasy.org/cgi-bin/prosite/nicedoc.pl?PDOC50020 says you find a conserved core; \n",
    "    how many do you need for a conserved core? align each one globally to find the conserved core. Does align_length find the conserved\n",
    "    core number? protein domains can be annotated for structural and functional similarity across species. \n",
    "    \n",
    "<img src=\"images/pa1_quiz_23.png\" >\n",
    "assuming domain is a subsequence of protein, find particular domain in protein. protein domains are conserved core:v\n",
    "    so this is global. \n",
    "<img src=\"images/pa1_quiz_24.png\" >\n",
    "wrong, this is done w local: albumin:582.0, ets:345.0, sh2:367.0, trypsin:367.0, ubx:306.0, albumin\n",
    "global: trypsin\n",
    "<img src=\"images/pa1_quiz_25.png\" >\n",
    "194 global trypsin:1227.0\n",
    "trypsin is enzyme to cleace proteins\n",
    "<img src=\"images/pa1_quiz_26.png\" >\n",
    "<img src=\"images/pa1_quiz_27.png\" >"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h6><Q4/h6>\n",
    "<img src=\"images/pa1_quiz_28.png\" >\n",
    "1578.0\n",
    "<img src=\"images/pa1_quiz_29.png\" >\n",
    "less, ancestor was much higher up between bacteria and people. What are the other differences between\n",
    "gaps or ...?\n",
    "<img src=\"images/pa1_quiz_30.png\" >\n",
    "<img src=\"images/pa1_quiz_31.png\" >\n",
    "<img src=\"images/pa1_quiz_32.png\" >\n",
    "<img src=\"images/pa1_quiz_33.png\" >\n",
    "Average number of gaps per alignment: 16.75\n",
    "Average length of gaps (this is the average of all averages): 1.47\n",
    "Average number of mismatches per alignment: 65.25\n",
    "open gap 5:\n",
    "Average number of gaps per alignment: 5.00\n",
    "Average length of gaps (this is the average of all averages): 2.20\n",
    "Average number of mismatches per alignment: 75.00\n",
    "open gap 0.5:\n",
    "Average number of gaps per alignment: 19.00\n",
    "Average length of gaps (this is the average of all averages): 1.35\n",
    "Average number of mismatches per alignment: 64.67\n",
    "    <h6>fewer gaps and more mismatches</h6>\n",
    "<img src=\"images/pa1_quiz_34.png\" >\n",
    "same mismatches, little less but more gaps\n",
    "<img src=\"images/pa1_quiz_36.png\" >\n",
    "<img src=\"images/pa1_quiz_37.png\" >\n",
    "PAM is based on gold standard manually aligned which perservs physical accuracy but does not mean it is optimal\n",
    "for alignment accuracy. Would expect PAM40 to be worse than PAM250 since there are less values and using teh default\n",
    "of match=1 and mismatch=-1 is clearly not accurate as given by the entries in blosum62. \n",
    "<img src=\"images/pa1_quiz_38.png\" >\n",
    "<img src=\"images/pa1_quiz_39.png\" >"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
