{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"11192019_01.png\">\n",
    "Simplest hypothesis was single gene creates single protein. Another step up in complexity is\n",
    "to think of phenotypes and how to break down the larger concept of an observable phenotype. \n",
    "\n",
    "monogenic: one gene causes observable problem like cystic fibrosis, if you have 2 bad copies of CFTR gene. \n",
    "    \n",
    "CFTR receptor lets salts and fluids flow in/out of cell. Most traits are polygenic. Where you have a continuous\n",
    "spectrum of observables and some set of genes associated with this. height is polygenic, multiple\n",
    "genes for hieght. MOst extreme is omigenic where most of genes lead to some characteristic like height. \n",
    "Wow. Thousands of genes will lead to a phenotype of say diabetes. \n",
    "\n",
    "\n",
    "\n",
    "<img src=\"11192019_02.png\">\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "benefits of networks, more data, and components are connected. interacting a protein to other proteins or components. \n",
    "nework representation can then use common tools to problems. Flexible. understand data drawbacks. for example data can \n",
    "be wrong in certain assays. have to understnd false positives and false negatives. also incomplete so have to underdsand\n",
    "this part. \n",
    "<img src=\"11192019_3.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Different types of networks, undirected=2 nodes connectd w/edge , directed=diretion on edge, cyclic; network w/cyccle, \n",
    "DAG: directed networks wo cycle, acyclic no cycle, cyclic has cycle, weighted: edge has quantative value, self-loops: node to iteself like\n",
    "in debruijn lecture. complete network: nodes connected w/all edges. homogeneous:1 node and edge types nonhomo: many node types\n",
    "and edge types. heteregenous network: bipartite graph, where there are 2 sets of nodes and set of edges conneting all nodes\n",
    "to each other, knowledge graph: eneities and directed edges and the edges qualify by some semantic meaning, different\n",
    "edge types are binds or relationships and nodes are entity. \n",
    "<img src=\"11192019_04.png\">\n",
    "<img src=\"11192019_05.png\">\n",
    "definition of graph, set of vertices and edges, each edge is a set of vertices. one representation is adjacency matrix another is adj list. \n",
    "sparsity is number of 0s; the most sparse is nothing, no edges and no nodes. Usually we mean there are N edges and dense \n",
    "approaching n squared edges. not hard definition. Sparse=use edge or edge list. of dense use adj matrix. \n",
    "<img src=\"11192019_06.png\">\n",
    "Local: subgraph can take out some nodes and edges from complete graph. anything is subgraph of graph. \n",
    "clique: is comlete subgraph. The whole neteowrk is s clique. ??? uncelar. waht is maximan clique? network motif:\n",
    "see a lot of little house shapes. maybe this means a regulatory mechanism. motif is a subgraph which occurs\n",
    "frequently enough it is important for your problem. \n",
    "<img src=\"11192019_07.png\">\n",
    "quantifying networks: \n",
    "conected components, 2 connected components\n",
    "    degee how many neighbors does node have? \n",
    "    if network is directed, talk about in indegrees and outdegrees. \n",
    "    look at degree distribution of network together using histogram. one things you see in bio network\n",
    "    is on log log sale, see a staircase linear in log log space. this is called a power law distribution. \n",
    "    if you see this then this is a scale free network. why did he make big deal about htis? this is small world. \n",
    "    or network is small diameter. For a network you can compute shortest path. YOu can comptue how to get between networks\n",
    "    in shortest way. diameter = longest shortest path. small world phenomana means there is root node where you can go \n",
    "    and transmit information efficiently. for bio network. small world is diameter is low. ?? not clear. \n",
    "    \n",
    "    \n",
    "<img src=\"11192019_08.png\">\n",
    "2) stable, if you take out node, nodes are vunlerable to targeteed attacks. \n",
    "clustering coefficient: refers to sparsity of node. \n",
    "Clustering coefficient between 0-1\n",
    "<img src=\"11192019_09.png\">\n",
    "bwtweeness: if you have to get to another node you have to go through a certain node. [0,1]\n",
    "how do you know if these stats are significant. if you do hypothesis you need a random model and compare vs. this.\n",
    "\n",
    "<img src=\"11192019_10.png\">\n",
    "network databases, molecular biology. \n",
    "1) regulatory nework. transcription factors and how genes are being transcribed for translation. a protein D \n",
    "which can inhibit the TF A from binding to DNA. D blocks A from notation. TransFAC has this info. \n",
    "2) metabolic network, representation of metabolites. how metabolies are alterted. metabolites are catalyzed by \n",
    "enzymes. a, b, care metabolines in diamond. The circle A and B are enzyme. We can model efect of enzymes on pathway.\n",
    "undirected weighted or unweighed graph. KEGG database. a,b,c are small molecules. \n",
    "\n",
    "<img src=\"11192019_11.png\">\n",
    "3) signalling netowrk. cell membraine with receptor. Have ligand that binds to receptor and hwen it binds teh receptor\n",
    "undergoes a conformational change which will activaet a proteein which will activate a TF and maybe activates a gene. \n",
    "unweighted. KEGG has this information. \n",
    "<img src=\"11192019_12.png\">\n",
    "4) protein protein interaction. historrcally about proteins physically binding together. The edge is proitens\n",
    "binding together. STRING database for protein protein. the original intention was physical pbinding. STRING contais\n",
    "experimental data, co-expression, text mining. and other smaller knowledge bases. \n",
    "5) funcitonal co-expression networks. model correlation of network. .9 and .8 spearman correlation. and .5 \n",
    "<img src=\"11192019_13.png\"> \n",
    "drugbank how to relate to target as birpartite network. ParmGKB database of drugs how it is processed by doby\n",
    "and genetic variants. \n",
    "<img src=\"11192019_14.png\">\n",
    "how to build your own database? 1) join the differenet networks togetner, \n",
    "2) crate network node by node. expression data w 4 genes, across different samples. want a network, use KNN which \n",
    "has most similar expression profile? directed network. individual genes and see how htey are related to each htoer\n",
    "3) edge by edge network construction. build Knowledge graph from literature, a PPI network. If I can qualify\n",
    "the network for protein protein interaction. If we can browse the papers and extract from papers and build a network\n",
    " from teh papers. will miss a lot of things but will have some information. Use NLP. \n",
    "<img src=\"11192019_15.png\">\n",
    "pipeline for doing this. NER to get terms or entities like SL2, EGFR, PLCgamma, \n",
    "Normalize:map the entities back to ontology\n",
    "exract relations: bind \n",
    "extract event: event is PLCGamma1 that sees binds and the EGFR is the receiver of hte action. exracted a triples\n",
    "    which is 2 nodes and edge. ONce you ahve big set of tribpes you can creat network. BMI 210. is about these ontologies~~\n",
    "    \n",
    "<img src=\"11192019_16.png\">\n",
    "task1: community detection, unsupervised clsutering, identify network communities. peripheral blood mononuclear cells,\n",
    "have scRNA data. want to knowif there are communities involved in RNAseq data? build network from RNASeq data!!\n",
    "really!! using KNN. apply louvain clustring!! modularity: how good is our clustering, quantifies density of links\n",
    "inside vs outside community. \n",
    "<img src=\"11192019_17.png\">\n",
    "algo: hierarchial clustering, initi one node=one community.2) for each i cancluate deltaQ or change in modularity\n",
    "    if I move i into cj. j is neignbor. Calculate deltaQ for all possible neighbors. 3) make a community level network\n",
    "    4) go to step 2 until convergence. \n",
    "<img src=\"11192019_18.png\">\n",
    "example of community level network. make super nodes which contain communities there is edge betwen AB but not\n",
    "between B and C. So create hierarchial community. Didnt have to specify k. Converges. \n",
    "<img src=\"11192019_19.png\">\n",
    "task2: node classification, give me functionally related nodes, guilt by association, iterative diffusion process\n",
    "input: list of query label genes. Konw a few genes want ot infer using guilt by association idea. output: labels. wamt\n",
    "        to infer what rest of genes do. \n",
    "<img src=\"11192019_20.png\">\n",
    "algorithm: \n",
    "1) init first iteration predictive labels based on labels you have. query genes have label 1 everything is 0. \n",
    "2) update: update with (1-lambda)y + lambdaSum(wijfj)\n",
    "weights from edges. calso calle lable propagation. Genemania implements this!!!\n",
    "\n",
    "\n",
    "<img src=\"11192019_21.png\">   \n",
    "task3: network alignment. 2 networks want to find correspondence between node and network. yeast network and human \n",
    "    gene netrowk. if we know the red lines are homologous proteins we are done. to do this munk. apply node embedding\n",
    "    try ot find mapping between nodes and some space. \n",
    "<img src=\"11192019_22.png\">\n",
    "red network is yeast, black is human, they line up in embedding space.representing nodes as vectors or embeddings. \n",
    "\n",
    "q) how do you build the embeddings? \n",
    "\n",
    "<img src=\"11192019_23.png\">\n",
    "link prediction: didnt talke about. 2 nodes shouhd have links between them. \n",
    "<img src=\"11192019_24.png\">\n",
    "\n",
    "<img src=\"11192019_25.png\">\n",
    "<img src=\"11192019_26.png\">\n",
    "<img src=\"11192019_27.png\">\n",
    "<img src=\"11192019_28.png\">\n",
    "<img src=\"11192019_29.png\">\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
