{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72,)\n",
      "['ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL'\n",
      " 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL'\n",
      " 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL'\n",
      " 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'ALL' 'AML' 'AML' 'AML' 'AML'\n",
      " 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML'\n",
      " 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML' 'AML']\n",
      "y_aml replace AML with 1 and ALL 0\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [nan nan]  tpr: [0. 1.]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n",
      "(71, 150) (1, 150) (71,) (1,)\n",
      "fpr: [0. 1.]  tpr: [nan nan]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:651: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/Users/dougchang/miniconda3/lib/python3.7/site-packages/sklearn/metrics/ranking.py:659: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "import pandas as pd\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "import numpy as np\n",
    "\n",
    "df = pd.read_csv(\"leukemia.csv\")\n",
    "#print(df.head())\n",
    "#print(df.columns)\n",
    "#X = df[\"Zyxin,\"PRG1 Proteoglycan 1, secretory granule\",CD33 CD33 antigen (differentiation antigen),DF D component of complement (adipsin),RNS2 Ribonuclease 2 (eosinophil-derived neurotoxin; EDN),CST3 Cystatin C (amyloid angiopathy and cerebral hemorrhage),APLP2 Amyloid beta (A4) precursor-like protein 2,\"GLUTATHIONE S-TRANSFERASE, MICROSOMAL\",CTSD Cathepsin D (lysosomal aspartyl protease),ATP6C Vacuolar H+ ATPase proton channel subunit,PROTEASOME IOTA CHAIN,LYN V-yes-1 Yamaguchi sarcoma viral related oncogene homolog,PPGB Protective protein for beta-galactosidase (galactosialidosis),HKR-T1,SPI1 Spleen focus forming virus (SFFV) proviral integration oncogene spi1,LYZ Lysozyme1,Lysozyme gene (EC 3.2.1.17),LYZ Lysozyme2,FAH Fumarylacetoacetate,RABAPTIN-5 protein,Leukotriene C4 synthase (LTC4S) gene,MYL1 Myosin light chain (alkali),ALDOA Aldolase A,\"ARHG Ras homolog gene family, member G (rho G)\",Liver mRNA for interferon-gamma inducing factor(IGIF),TCF3 Transcription factor 3 (E2A immunoglobulin enhancer binding factors E12/E47)1,MPO Myeloperoxidase,MB-1 gene,\"FTL Ferritin, light polypeptide\",Azurocidin gene,TIMP2 Tissue inhibitor of metalloproteinase 2,Nucleoside-diphosphate kinase,Macmarcks,ALDR1 Aldehyde reductase 1 (low Km aldose reductase),ME491  gene extracted from H.sapiens gene for Me491/CD63 antigen,GPX1 Glutathione peroxidase 1,LAMP2 Lysosome-associated membrane protein 2 {alternative products},\"C-myb gene extracted from Human (c-myb) gene, complete primary cds, and five complete alternatively spliced cds\",\"Quiescin (Q6) mRNA, partial cds\",SELL Leukocyte adhesion protein beta subunit2,CCND3 Cyclin D3,TCF3 Transcription factor 3 (E2A immunoglobulin enhancer binding factors E12/E47)2,\"KAI1 Kangai 1 (suppression of tumorigenicity 6, prostate; CD82 antigen (R2 leukocyte antigen, antigen detected by monoclonal and antibody IA4))\",\"ITGAX Integrin, alpha X (antigen CD11C (p150), alpha polypeptide)\",\"PFC Properdin P factor, complement\",PROBABLE G PROTEIN-COUPLED RECEPTOR LCR1 HOMOLOG,Transcriptional activator hSNF2b1,INTERLEUKIN-8 PRECURSOR,Oncoprotein 18 (Op18) gene,\"Clone 22 mRNA, alternative splice variant alpha-1\",CYSTATIN A,14-3-3 PROTEIN TAU,Hunc18b2,PTH2 parathyroid hormone receptor mRNA,MSH2 DNA repair protein MSH2,MANB Mannosidase alpha-B (lysosomal),KIAA0022 gene,\"SPTAN1 Spectrin, alpha, non-erythrocytic 1 (alpha-fodrin)\",VIL2 Villin 2 (ezrin),ALDH7 Aldehyde dehydrogenase 7,\"ZNF91 Zinc finger protein 91 (HPF7, HTF10)\",Keratin 10 type I intermediate filament (KRT10) mRNA,\"GB DEF = Topoisomerase type II (Topo II) mRNA, partial cds\",INTERFERON GAMMA UP-REGULATED I-5111 PROTEIN PRECURSOR,CDC10 Cell division cycle 10 (homologous to CDC10 of S. cerevisiae,26-kDa cell surface protein TAPA-1 mRNA,TOP2B Topoisomerase (DNA) II beta (180kD),SMT3B protein,\"Calcineurin A catalytic subunit [human, testis, mRNA, 2134 nt]\",\"PRKCD Protein kinase C, delta\",Inducible protein mRNA,\"GTF2E2 General transcription factor TFIIE beta subunit, 34 kD\",NADH:ubiquinone oxidoreductase subunit B13 (B13) mRNA,Thymopoietin beta mRNA,\"TCF12 Transcription factor 12 (HTF4, helix-loop-helix transcription factors 4)\",LEUKOCYTE ELASTASE INHIBITOR,GRN Granulin,CRYZ Crystallin zeta (quinone reductase),PERIPHERAL-TYPE BENZODIAZEPINE RECEPTOR,PCBD 6-pyruvoyl-tetrahydropterin synthase/dimerization cofactor of hepatocyte nuclear factor 1 alpha (TCF1),\"Heat Shock Protein, 70 Kda (Gb:Y00371)\",ERC-55 mRNA,Terminal transferase mRNA,PLECKSTRIN,\"RHD Rhesus blood group, D antigen\",Transcriptional activator hSNF2b2,\"CHRNA7 Cholinergic receptor, nicotinic, alpha polypeptide 7\",\"GB DEF = Retinoblastoma susceptibility protein (RB1) gene, with a 3 bp deletion in exon 22 (L11910 bases 161855-162161)\",IL7R Interleukin 7 receptor,PHOSPHATIDYLINOSITOL,PGD Phosphogluconate dehydrogenase,GB DEF = GTP-binding protein (RAB3B) mRNA,CAB3b mRNA for calcium channel beta3 subunit,\"CD36 CD36 antigen (collagen type I receptor, thrombospondin receptor)\",\"KIAA0128 gene, partial cds\",SELL Leukocyte adhesion protein beta subunit1,Nuclear Factor Nf-Il6,Phosphotyrosine independent ligand p62 for the Lck SH2 domain mRNA,AKT1 V-akt murine thymoma viral oncogene homolog 1,LRPAP1 Low density lipoprotein-related protein-associated protein 1 (alpha-2-macroglobulin receptor-associated protein 1,SNRPN Small nuclear ribonucleoprotein polypeptide N,Amyloid precursor protein-binding protein 1 mRNA,DCK Deoxycytidine kinase,ADPRT ADP-ribosyltransferase (NAD+; poly (ADP-ribose) polymerase),GB DEF = P85 beta subunit of phosphatidyl-inositol-3-kinase,Clone 23721 mRNA sequence,Thymopoietin (TMPO) gene,GB DEF = DNA for cellular retinol binding protein (CRBP) exons 3 and 4,\"ACADM Acyl-Coenzyme A dehydrogenase, C-4 to C-12 straight chain\",LMNA Lamin A,\"LMP2 gene extracted from H.sapiens genes TAP1, TAP2, LMP2, LMP7 and DOB\",\"SERINE/THREONINE PROTEIN PHOSPHATASE 2B CATALYTIC SUBUNIT, BETA ISOFORM\",GB DEF = Homeodomain protein HoxA9 mRNA,CSF3R Colony stimulating factor 3 receptor (granulocyte),Hepatocyte growth factor-like protein gene,Tetracycline transporter-like protein mRNA,\"Putative enterocyte differentiation promoting factor mRNA, partial cds\",\"PI Protease inhibitor 1 (anti-elastase), alpha-1-antitrypsin\",Epb72 gene exon 1,Interleukin 8 (IL8) gene,GB DEF = Myosin-IE,Activin type II receptor,Skeletal muscle LIM-protein SLIM1 mRNA,TALDO Transaldolase,Transmembrane protein,HMG1 High-mobility group (nonhistone chromosomal) proteinleukemia_type\"]\n",
    "y = df[\"leukemia_type\"] \n",
    "X=df.loc[:, df.columns != 'leukemia_type']\n",
    "y_np = y.to_numpy()\n",
    "print(y_np.shape)\n",
    "print(y_np)\n",
    "y_all = np.where(y_np==\"ALL\", 1, y_np) \n",
    "y_all= np.where(y_all==\"AML\",0,y_all)\n",
    "y_all=y_all.astype('int')\n",
    "#print(\"y_all replace ALL with 1 and AML 0\")\n",
    "#print(y_all)\n",
    "y_aml=np.where(y_np==\"ALL\", 0, y_np)\n",
    "y_aml=np.where(y_aml==\"AML\",0,y_aml)\n",
    "print(\"y_aml replace AML with 1 and ALL 0\")\n",
    "#print(y_aml)\n",
    "knn = KNeighborsClassifier(n_neighbors=5)\n",
    "\n",
    "loo = LeaveOneOut()\n",
    "for train_index, test_index in loo.split(X):\n",
    "    X_train, X_test = X.to_numpy()[train_index], X.to_numpy()[test_index]\n",
    "    #y_train, y_test = y.to_numpy()[train_index], y.to_numpy()[test_index]\n",
    "    y_train, y_test = y_all[train_index], y_all[test_index]\n",
    "    print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)\n",
    "    probas_ = knn.fit(X_train, y_train).predict(X_test)\n",
    "    # Compute ROC curve and area the curve\n",
    "    #print(type(probas_),probas_)\n",
    "    #print(y_test)\n",
    "    fpr, tpr, thresholds = roc_curve(y_test, probas_,pos_label=1)\n",
    "    print(\"fpr:\",fpr,' tpr:',tpr)\n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====\n",
      "<class 'sklearn.model_selection._split.LeaveOneOut'>\n",
      "====\n",
      "(72, 150) (72,)\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "tp_ALL: 44  tn_ALL: 21  fp_ALL: 7  fn_ALL: 0 tpr_all: 1.0 fpr_all: 0.25\n",
      "tp_AML: 21  tn_AML: 44  fp_AML: 0  fn_AML: 7 tpr_aml: 0.75 fpr_aml: 0.0\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import LeaveOneOut\n",
    "loo = LeaveOneOut()\n",
    "#loo.get_n_splits(X,y)\n",
    "print('====')\n",
    "print(type(loo))\n",
    "print('====')\n",
    "print(X.shape,y.shape)\n",
    "tp_ALL=0\n",
    "tn_ALL=0\n",
    "fp_ALL=0\n",
    "fn_ALL=0\n",
    "tp_AML=0\n",
    "tn_AML=0\n",
    "fp_AML=0\n",
    "fn_AML=0\n",
    "\n",
    "for train_index, test_index in loo.split(X):\n",
    "    #print(\"TRAIN:\", train_index, \"TEST:\", test_index)\n",
    "    X_train, X_test = X.to_numpy()[train_index], X.to_numpy()[test_index]\n",
    "    y_train, y_test = y.to_numpy()[train_index], y.to_numpy()[test_index]\n",
    "    #print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)\n",
    "    knn = KNeighborsClassifier(n_neighbors=5)\n",
    "    knn.fit(X_train, y_train) \n",
    "    res = knn.predict(X_test)\n",
    "    print(y.to_numpy()[test_index],res)\n",
    "    if y.to_numpy()[test_index]==\"ALL\" and res==\"ALL\":\n",
    "        tp_ALL +=1\n",
    "        tn_AML+=1\n",
    "    elif(y.to_numpy()[test_index]==\"ALL\" and res==\"AML\"):\n",
    "        fn_ALL+=1\n",
    "        fp_AML+=1\n",
    "    elif(y.to_numpy()[test_index]==\"AML\" and res==\"AML\"):\n",
    "        tn_ALL+=1\n",
    "        tp_AML+=1\n",
    "    elif(y.to_numpy()[test_index]==\"AML\" and res==\"ALL\"):\n",
    "        fp_ALL+=1\n",
    "        fn_AML+=1\n",
    "    else:\n",
    "        print('not see')\n",
    "#print(tp,tn,fp,fn)\n",
    "print(\"tp_ALL:\",tp_ALL,\" tn_ALL:\",tn_ALL,\" fp_ALL:\",fp_ALL,\" fn_ALL:\",fn_ALL, \"tpr_all:\",(tp_ALL)/(tp_ALL+fn_ALL),\"fpr_all:\",fp_ALL/(fp_ALL+tn_ALL))\n",
    "print(\"tp_AML:\",tp_AML,\" tn_AML:\",tn_AML,\" fp_AML:\",fp_AML,\" fn_AML:\",fn_AML, \"tpr_aml:\",(tp_AML)/(tp_AML+fn_AML),\"fpr_aml:\",fp_AML/(fp_AML+tn_AML))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====\n",
      "<class 'sklearn.model_selection._split.LeaveOneOut'>\n",
      "====\n",
      "(72, 150) (72,)\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['ALL'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['AML']\n",
      "['AML'] ['ALL']\n",
      "tp_ALL: 44  tn_ALL: 24  fp_ALL: 0  fn_ALL: 4 tpr_all: 0.9166666666666666 fpr_all: 0.0\n",
      "tp_AML: 24  tn_AML: 44  fp_AML: 0  fn_AML: 4 tpr_aml: 0.8571428571428571 fpr_aml: 0.0\n"
     ]
    }
   ],
   "source": [
    "from sklearn.naive_bayes import GaussianNB\n",
    "\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "loo = LeaveOneOut()\n",
    "loo.get_n_splits(X,y)\n",
    "print('====')\n",
    "print(type(loo))\n",
    "print('====')\n",
    "print(X.shape,y.shape)\n",
    "tp_ALL=0\n",
    "tn_ALL=0\n",
    "fp_ALL=0\n",
    "fn_ALL=0\n",
    "tp_AML=0\n",
    "tn_AML=0\n",
    "fp_AML=0\n",
    "fn_AML=0\n",
    "\n",
    "for train_index, test_index in loo.split(X):\n",
    "    #print(\"TRAIN:\", train_index, \"TEST:\", test_index)\n",
    "    X_train, X_test = X.to_numpy()[train_index], X.to_numpy()[test_index]\n",
    "    y_train, y_test = y.to_numpy()[train_index], y.to_numpy()[test_index]\n",
    "    #print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)\n",
    "    nb = GaussianNB()\n",
    "    nb.fit(X_train, y_train) \n",
    "    res = nb.predict(X_test)\n",
    "    print(y.to_numpy()[test_index],res)\n",
    "    if y.to_numpy()[test_index]==\"ALL\" and res==\"ALL\":\n",
    "        tp_ALL +=1\n",
    "        tn_AML+=1\n",
    "    elif(y.to_numpy()[test_index]==\"ALL\" and res==\"AML\"):\n",
    "        fn_ALL+=1\n",
    "        fp_AML+=1\n",
    "    elif(y.to_numpy()[test_index]==\"AML\" and res==\"AML\"):\n",
    "        tn_ALL+=1\n",
    "        tp_AML+=1\n",
    "    elif(y.to_numpy()[test_index]==\"AML\" and res==\"ALL\"):\n",
    "        fn_ALL+=1\n",
    "        fn_AML+=1\n",
    "    else:\n",
    "        print('not see')\n",
    "print(\"tp_ALL:\",tp_ALL,\" tn_ALL:\",tn_ALL,\" fp_ALL:\",fp_ALL,\" fn_ALL:\",fn_ALL, \"tpr_all:\",(tp_ALL)/(tp_ALL+fn_ALL),\"fpr_all:\",fp_ALL/(fp_ALL+tn_ALL))\n",
    "print(\"tp_AML:\",tp_AML,\" tn_AML:\",tn_AML,\" fp_AML:\",fp_AML,\" fn_AML:\",fn_AML, \"tpr_aml:\",(tp_AML)/(tp_AML+fn_AML),\"fpr_aml:\",fp_AML/(fp_AML+tn_AML))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====\n",
      "<class 'sklearn.model_selection._split.LeaveOneOut'>\n",
      "====\n",
      "(4, 2) (4, 1)\n",
      "[1 2 3] [0]\n",
      "[[1 1]]\n",
      "[0 2 3] [1]\n",
      "[[2 2]]\n",
      "[0 1 3] [2]\n",
      "[[3 3]]\n",
      "[0 1 2] [3]\n",
      "[[4 4]]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import LeaveOneOut\n",
    "import numpy as np\n",
    "loo = LeaveOneOut()\n",
    "X=np.array([[1,1],[2,2],[3,3],[4,4]])\n",
    "y=np.array([[\"a\"],[\"b\"],[\"c\"],[\"d\"]])\n",
    "loo.get_n_splits(X,y)\n",
    "print('====')\n",
    "print(type(loo))\n",
    "print('====')\n",
    "print(X.shape,y.shape)\n",
    "tp=0\n",
    "tn=0\n",
    "fp=0\n",
    "fn=0\n",
    "\n",
    "for train_index, test_index in loo.split(X):\n",
    "    print(train_index,test_index)\n",
    "    X_train, X_test = X[train_index], X[test_index]\n",
    "    y_train, y_test = y[train_index], y[test_index]\n",
    "    #print(X_train,y_train)\n",
    "    print(X_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"images/pa2_1.png\">\n",
    "<img src=\"images/pa2_2.png\">\n",
    "<img src=\"images/pa2_3.png\">\n",
    "<img src=\"images/pa2_4.png\">\n",
    "<img src=\"images/pa2_5.png\">\n",
    "<img src=\"images/pa2_6.png\">\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h6>Q2</h6>\n",
    "<img src=\"images/pa2_7.png\">\n",
    "<img src=\"images/pa2_8.png\">\n",
    "<img src=\"images/pa2_10.png\">\n",
    "<img src=\"images/pa2_11.png\">\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Diploid=2 sets of chromosomes; 1 set from each parent. 2 copies of same gene; one from each parent for diploid. \n",
    "<img src=\"images/pa2_12.png\">\n",
    "Allele pair of same gene in 2 chromosomes. Allele from 2 parents have SNP vaiartions. \n",
    "<img src=\"images/pa2_13.png\">\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4690620/ 81.7m\n",
    "<img src=\"images/pa2_14.png\">\n",
    "translocation; indels\n",
    "recombination,copy number variation , gene deletion, inversion,\n",
    "<img src=\"images/pa2_18.png\">\n",
    "50% penetrance is 50% of population w/particular Allele or gene variant will develop trait or disease associated w/gene variation. \n",
    "95% penetrance, 95% of popuLATION will express trait. nasty. \n",
    "<img src=\"images/pa2_19.png\">\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A haplotype is a specific combination of SNPs on the same chromosome. SNPs can be determined by comapring genotype information from 2 paretns. IE allele ATCT and ATCA. if last NT is C then we have SNP.   \n",
    "Allele A=.25% on chromosome, B=.25% then the probability\n",
    "of the homologous combined chromosome having AB is .50; pA*Pb. these are independent. and linkage disequilibrimwhen the alleles are too close togetreh\n",
    "and they arent far enough on the chromosome for homologous recombination. 2 alleles close together can be a hapolotype. can predice LD frequency given positions of alleles on chromosomes. \n",
    "<img src=\"images/pa2_21.png\">\n",
    "1)plink --file hapmap1 --make-bed --out hapmap1; make binary ped file for plink which reduces storage. command is \"make-bed\"\n",
    "2)plink --bfile hapmap1 --missing --out miss_stat; missing data statistics, clean data. command is \"missing\"\n",
    "3)plink --bfile hapmap1 --assoc --out as1; perform association analysis between alleles and disease, command is \"assoc\"\n",
    "4)plink --bfile hapmap1 --cluster --mc 2 --ppc 0.05 --out str1; cluster analysis, if individuals belong to same population\n",
    "5)plink --bfile hapmap1 --model --snp rs2222162 --out mod1; chochran-armitage test\n",
    "<img src=\"images/pa2_22.png\">\n",
    "<img src=\"images/pa2_23.png\">\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
